### PR TITLE
Changed XMA plugin code with XRT native infra

### PR DESCRIPTION
**Pull request by vboggara-xilinx**
#### Problem solved by the commit
Moved XclBufferHandles to xrt::bo in XMA plugin code
#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Its PR related work, https://jira.xilinx.com/browse/VITIS-3278
#### Risks (if any) associated the changes in the commit
Low Risk
#### What has been tested and how, request additional testing if necessary
Ran XMA sprite testcases locally with this change, all are passing without any issue
#### Documentation impact (if any)**
N/A

### CODE REVIEW

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 1]
```diff
@@ -375,12 +310,7 @@ xma_plg_buffer_write(XmaSession s_handle,
         return XMA_SUCCESS;
     }
 
-    rc = xclSyncBO(b_obj_priv->dev_handle, b_obj_priv->boHandle, XCL_BO_SYNC_BO_TO_DEVICE, size, offset);
-    if (rc != 0) {
-        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_write failed. dev_index: %d. xclSyncBO failed. Check device logs for more info. Error: %d", b_obj_priv->dev_index, rc);
-        return XMA_ERROR;
-    }
-
+    b_obj_priv->xrt_bo.sync(XCL_BO_SYNC_BO_TO_DEVICE, size, offset);
```
Comment (vboggara-xilinx) [inline src/xma/src/xmaplugin/xmaplugin.cpp:313]: @stsoe , currently we are relying on system error thrown by xrt::bo::sync, do we need to explicitly handle this and throw XMA_ERROR?

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 2]
```diff
@@ -35,24 +37,18 @@ static_assert(sizeof(XmaCmdState) <= sizeof(int32_t), "XmaCmdState size must be
...
-
-    struct xclBOProperties bop = {0};
-    if (xclGetBOProperties(dev_handle, b_obj_handle, &bop) != 0) {
-        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_alloc failed to get BO properties");
-        xclFreeBO(dev_handle, b_obj_handle);
-        return XMA_ERROR;
-    }
-    b_obj.paddr = bop.paddr;
-
+        xrt_bo_obj = xrt::bo(dev_handle, size, ddr_bank);
```
Comment (vboggara-xilinx) [inline src/xma/src/xmaplugin/xmaplugin.cpp:46]: @stsoe , do we need to handle this XMA_ERROR?

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 3]
```diff
@@ -21,12 +21,14 @@
...
 #include "app/xma_utils.hpp"
 #include "lib/xma_utils.hpp"
+#include "core/common/api/bo.h"
 
 #include <cstdio>
 #include <iostream>
 #include <cstring>
 #include <thread>
 #include <chrono>
+#include <algorithm>
```
Comment (keryell) [inline src/xma/src/xmaplugin/xmaplugin.cpp:31]: At least sort the includes you add...

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 4]
```diff
@@ -79,186 +74,130 @@ xma_plg_buffer_alloc(XmaSession s_handle, size_t size, bool device_only_buffer,
...
+
+XmaBufferObj  create_xma_buffer_object(XmaSession s_handle, size_t size, bool device_only_buffer, uint32_t ddr_bank ,int32_t* return_code) {
+    XmaBufferObj b_obj;
+    XmaBufferObj b_obj_error = create_error_bo(); 
     b_obj.data = nullptr;
     //b_obj.ref_cnt = 0;
     b_obj.user_ptr = nullptr;
     b_obj.device_only_buffer = false;
-    b_obj.private_do_not_touch = nullptr;
+    b_obj.private_do_not_touch = nullptr;
```
Comment (keryell) [inline src/xma/src/xmaplugin/xmaplugin.cpp:87]: Why?

Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp:87]: sees some formatting issue with my editor, will remove extra space.

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 3 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp:31]: ok, will sort them.

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 5]
```diff
@@ -308,7 +229,7 @@ int32_t xma_check_device_buffer(XmaBufferObj *b_obj) {
         xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_check_device_buffer failed. XMABufferObj failed allocation");
         return XMA_ERROR;
     }
-    if (b_obj_priv->dev_index < 0 || b_obj_priv->bank_index < 0 || b_obj_priv->size <= 0) {
+    if (b_obj->dev_index < 0 || xrt_core::bo::group_id(b_obj_priv->xrt_bo) < 0 || b_obj_priv->xrt_bo.size() <= 0) {
```
Comment (stsoe) [inline src/xma/src/xmaplugin/xmaplugin.cpp:232]: This is déjà vu, didn't we see this same code in xmabuffer.cpp?

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 6]
```diff
@@ -353,7 +268,7 @@ xma_plg_buffer_write(XmaSession s_handle,
                      size_t           size,
                      size_t           offset)
 {
-    int32_t rc;
+    //int32_t rc;
```
Comment (stsoe) [inline src/xma/src/xmaplugin/xmaplugin.cpp:271]: remove?

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 7]
```diff
@@ -390,7 +301,7 @@ xma_plg_buffer_read(XmaSession s_handle,
                     size_t           size,
                     size_t           offset)
 {
-    int32_t rc;
+    //int32_t rc;
```
Comment (stsoe) [inline src/xma/src/xmaplugin/xmaplugin.cpp:304]: remove?

**Review by stsoe — CHANGES_REQUESTED** [review]
This is very good.   Nice first step in simplifying XMA.  

We definitely want this evolutionary path towards the end-goal and reviewers should understand that the changes we are making are not the final goal, but steps in the right direction. 

My comments are just small things.

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 8]
```diff
@@ -35,24 +39,18 @@ static_assert(sizeof(XmaCmdState) <= sizeof(int32_t), "XmaCmdState size must be
...
-    }
-    b_obj.paddr = bop.paddr;
-
+        xrt_bo_obj = xrt::bo(dev_handle, size, ddr_bank);
+    }        
+    b_obj.paddr = xrt_bo_obj.address();
     if (!device_only_buffer) {
-        b_obj.data = (uint8_t*) xclMapBO(dev_handle, b_obj_handle, true);
+        b_obj.data = xrt_bo_obj.map<uint8_t*>();
+	    std::fill(b_obj.data, b_obj.data + size, 0);
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp:53]: Yes soren, its needed. we are seeing some testcases failing without this code.

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 1 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: added try catch block and returning XMA_ERROR when exception caught.

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 2 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: added try catch block and returning XMA_ERROR when exception caught.

File: src/runtime_src/core/common/api/xrt_bo.cpp [hunk 9]
```diff
@@ -892,6 +892,16 @@ group_id(const xrt::bo& bo)
   return bo.get_handle()->get_group_id();
 }
 
+xclDeviceHandle
+device_handle(const xrt::bo& bo) {
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/api/xrt_bo.cpp]: comments addressed.

File: src/runtime_src/core/common/api/xrt_bo.cpp [hunk 10]
```diff
@@ -892,6 +892,16 @@ group_id(const xrt::bo& bo)
...
+    return bo.get_handle()->get_device()->get_device_handle();
+}
+
+xrt::bo::flags
+get_flags(const xrt::bo& bo) {
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/runtime_src/core/common/api/xrt_bo.cpp]: comments addressed.

File: src/xma/include/lib/xmahw_lib.h [hunk 11]
```diff
@@ -224,25 +225,13 @@ typedef struct XmaHwSessionPrivate
...
   XmaBufferObjPrivate() {
-   dummy = NULL;
-   size = 0;
-   bank_index = -1;
-   dev_index = -1;
-   ref_cnt = 0;
-   dev_handle = NULL;
-   device_only_buffer = false;
-   boHandle = 0;
+   dummy = NULL;
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/include/lib/xmahw_lib.h]: comments addressed.

File: src/xma/src/xmaapi/xmabuffer.cpp [hunk 12]
```diff
@@ -144,15 +145,15 @@ int32_t xma_check_device_buffer(XmaBufferObj *b_obj) {
...
     }
     if (b_obj_priv->dummy != (void*)(((uint64_t)b_obj_priv) | signature)) {
         xma_logmsg(XMA_ERROR_LOG, XMA_BUFFER_MOD, "xma_check_device_buffer failed. XMABufferObj is corrupted.\n");
         return XMA_ERROR;
     }
-    if (!b_obj_priv->dev_handle) {
+    if (!xrt_core::bo::device_handle(b_obj_priv->xrt_bo)) {
         xma_logmsg(XMA_ERROR_LOG, XMA_BUFFER_MOD, "xma_check_device_buffer failed. XMABufferObj is corrupted.\n");
         return XMA_ERROR;
     }
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaapi/xmabuffer.cpp]: Removed this check.

File: src/xma/src/xmaapi/xmabuffer.cpp [hunk 13]
```diff
@@ -144,15 +145,15 @@ int32_t xma_check_device_buffer(XmaBufferObj *b_obj) {
         xma_logmsg(XMA_ERROR_LOG, XMA_BUFFER_MOD, "xma_check_device_buffer failed. XMABufferObj failed allocation\n");
         return XMA_ERROR;
     }
-    if (b_obj_priv->dev_index < 0 || b_obj_priv->bank_index < 0 || b_obj_priv->size <= 0) {
+    if (b_obj->dev_index < 0 || xrt_core::bo::group_id(b_obj_priv->xrt_bo) < 0 || b_obj_priv->xrt_bo.size() <= 0) {
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaapi/xmabuffer.cpp]: comments addressed.

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 14]
```diff
@@ -35,24 +39,18 @@ static_assert(sizeof(XmaCmdState) <= sizeof(int32_t), "XmaCmdState size must be
 extern XmaSingleton *g_xma_singleton;
 
 int32_t create_bo(xclDeviceHandle dev_handle, XmaBufferObj& b_obj, uint32_t size, uint32_t ddr_bank, 
-    bool device_only_buffer, xclBufferHandle& b_obj_handle) {
+    bool device_only_buffer, xrt::bo& xrt_bo_obj) {
     if (device_only_buffer) {
-        b_obj_handle = xclAllocBO(dev_handle, size, 0, XCL_BO_FLAGS_DEV_ONLY | ddr_bank);
+        xrt::bo::flags flags = xrt::bo::flags::device_only;
+        xrt_bo_obj = xrt::bo(dev_handle, size, flags, ddr_bank);
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: comments addressed.

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 15]
```diff
@@ -67,9 +65,8 @@ void cmd_obj_default(XmaCUCmdObj& cmd_obj) {
 }
 
 XmaBufferObj
-xma_plg_buffer_alloc(XmaSession s_handle, size_t size, bool device_only_buffer, int32_t* return_code)
+create_error_bo()
 {
-    XmaBufferObj b_obj;
     XmaBufferObj b_obj_error;
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp:77]: No Soren, we cant initialize them in XmaBufferObj  structure as this structure exposed to out by using extern c

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 16]
```diff
@@ -79,186 +76,129 @@ xma_plg_buffer_alloc(XmaSession s_handle, size_t size, bool device_only_buffer,
...
+}
+
+XmaBufferObj  create_xma_buffer_object(XmaSession s_handle, size_t size, bool device_only_buffer, uint32_t ddr_bank ,int32_t* return_code) {
+    XmaBufferObj b_obj;
+    XmaBufferObj b_obj_error = create_error_bo(); 
     b_obj.data = nullptr;
-    //b_obj.ref_cnt = 0;
     b_obj.user_ptr = nullptr;
     b_obj.device_only_buffer = false;
     b_obj.private_do_not_touch = nullptr;
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp:95]: Needed here. above comment applicable here.

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 17]
```diff
@@ -79,186 +76,129 @@ xma_plg_buffer_alloc(XmaSession s_handle, size_t size, bool device_only_buffer,
...
+        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_alloc_ddr failed. XMASession is corrupted.");
         if (return_code) *return_code = XMA_ERROR;
         return b_obj_error;
     }
-    XmaHwSessionPrivate *priv1 = (XmaHwSessionPrivate*) s_handle.hw_session.private_do_not_use;
+    XmaHwSessionPrivate* priv1 = (XmaHwSessionPrivate*)s_handle.hw_session.private_do_not_use;
     xclDeviceHandle dev_handle = priv1->dev_handle;
-    uint32_t ddr_bank = s_handle.hw_session.bank_index;
+    
+    //uint32_t ddr_bank = s_handle.hw_session.bank_index;
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: removed

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 18]
```diff
@@ -79,186 +76,129 @@ xma_plg_buffer_alloc(XmaSession s_handle, size_t size, bool device_only_buffer,
...
-        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_alloc: libxmaplugin can not be used without loading libxmaapi");
+        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma plugin buffer allocation can not be used for this XMASession type");
         if (return_code) *return_code = XMA_ERROR;
         return b_obj_error;
     }
-
-    xclBufferHandle b_obj_handle = 0;
+    
+    xrt::bo b_obj_handle;
     if (create_bo(dev_handle, b_obj, size, ddr_bank, device_only_buffer, b_obj_handle) != XMA_SUCCESS) {
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp:116]: comments addressed.

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 19]
```diff
@@ -79,186 +76,129 @@ xma_plg_buffer_alloc(XmaSession s_handle, size_t size, bool device_only_buffer,
...
-
+    XmaBufferObj b_obj_error = create_error_bo();
     if (xma_core::utils::check_xma_session(s_handle) != XMA_SUCCESS) {
         xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_alloc_ddr failed. XMASession is corrupted.");
         if (return_code) *return_code = XMA_ERROR;
         return b_obj_error;
     }
-    XmaHwSessionPrivate *priv1 = (XmaHwSessionPrivate*) s_handle.hw_session.private_do_not_use;
-    xclDeviceHandle dev_handle = priv1->dev_handle;
+    XmaHwSessionPrivate* priv1 = (XmaHwSessionPrivate*)s_handle.hw_session.private_do_not_use;
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: comments addressed.

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 5 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: moved this code to utils area.

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 6 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: comments addressed.

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 7 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: comments addressed.

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 20]
```diff
@@ -412,20 +302,19 @@ xma_plg_buffer_read(XmaSession s_handle,
...
-    {
-        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_read failed. dev_index: %d. xclSyncBO failed. Check device logs for more info. Error: %d", b_obj_priv->dev_index, rc);
-        return XMA_ERROR;
+    try {
+        b_obj_priv->xrt_bo.sync(XCL_BO_SYNC_BO_FROM_DEVICE, size, offset);
+        return XMA_SUCCESS;
     }
-
-    return XMA_SUCCESS;
+    catch (const xrt_core::system_error& ex) {
```
Comment (stsoe) [inline src/xma/src/xmaplugin/xmaplugin.cpp:309]: don't name the argument if it is not used, so in this case just do
`catch (const xrt_core::system_error&) {`

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 21]
```diff
@@ -412,20 +302,19 @@ xma_plg_buffer_read(XmaSession s_handle,
...
+        xma_logmsg(XMA_ERROR_LOG, XMAPLUGIN_MOD, "xma_plg_buffer_read failed. xclSyncBO failed. Check device logs for more info.");
+        return XMA_ERROR;
+    } 
 }
 
 int32_t xma_plg_execbo_avail_get(XmaSession s_handle)
 {
-    XmaHwSessionPrivate *priv1 = (XmaHwSessionPrivate*) s_handle.hw_session.private_do_not_use;
+    XmaHwSessionPrivate *priv1 = reinterpret_cast<XmaHwSessionPrivate*>(s_handle.hw_session.private_do_not_use);
```
Comment (stsoe) [inline src/xma/src/xmaplugin/xmaplugin.cpp:317]: Avoid repeating types.  Use auto instead.
`auto priv1 = reinterpret_cast<...>`

File: src/xma/src/xmaplugin/xmaplugin.cpp [hunk 22]
```diff
@@ -453,7 +342,7 @@ int32_t xma_plg_execbo_avail_get(XmaSession s_handle)
 
 int32_t xma_plg_execbo_avail_get2(XmaSession s_handle)
 {
-    XmaHwSessionPrivate *priv1 = (XmaHwSessionPrivate*) s_handle.hw_session.private_do_not_use;
+    XmaHwSessionPrivate *priv1 = reinterpret_cast<XmaHwSessionPrivate*>(s_handle.hw_session.private_do_not_use);
```
Comment (stsoe) [inline src/xma/src/xmaplugin/xmaplugin.cpp:345]: auto here and below

**Review by stsoe — CHANGES_REQUESTED** [review]
Looks good.  Just address these new comments I added, and then we will merge after next build.

File: src/xma/src/xmaapi/xma_utils.cpp [hunk 23]
```diff
@@ -693,6 +694,28 @@ int32_t check_all_execbo(XmaSession s_handle) {
     return XMA_SUCCESS;
 }
 
+int32_t xma_check_device_buffer(XmaBufferObj* b_obj) {
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaapi/xma_utils.cpp]: comments addressed

File: src/xma/src/xmaapi/xma_utils.cpp [hunk 24]
```diff
@@ -693,6 +694,28 @@ int32_t check_all_execbo(XmaSession s_handle) {
...
+    if (!b_obj) {
+        xma_logmsg(XMA_ERROR_LOG, XMAUTILS_MOD, "xma_check_device_buffer failed. XMABufferObj failed allocation\n");
+        return XMA_ERROR;
+    }
+
+    XmaBufferObjPrivate* b_obj_priv = reinterpret_cast<XmaBufferObjPrivate*>(b_obj->private_do_not_touch);
```
[the remark this answers is not in the document]
Comment (vboggara-xilinx) [reply src/xma/src/xmaapi/xma_utils.cpp]: comments addressed

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 20 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: comments addressed

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 21 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: comments addressed

File: src/xma/src/xmaplugin/xmaplugin.cpp (hunk 22 above)
Comment (vboggara-xilinx) [reply src/xma/src/xmaplugin/xmaplugin.cpp]: comments addressed

**Comment by vboggara-xilinx** [comment]
@stsoe , addressed your comments and submitted latest PR.

File: src/xma/src/xmaapi/xma_utils.cpp [hunk 25]
```diff
@@ -693,6 +694,28 @@ int32_t check_all_execbo(XmaSession s_handle) {
     return XMA_SUCCESS;
 }
 
+int32_t xma_check_device_buffer(const XmaBufferObj* b_obj) {
+    if (!b_obj) {
+        xma_logmsg(XMA_ERROR_LOG, XMAUTILS_MOD, "xma_check_device_buffer failed. XMABufferObj failed allocation\n");
+        return XMA_ERROR;
+    }
+
+    auto b_obj_priv = reinterpret_cast<XmaBufferObjPrivate*>(b_obj->private_do_not_touch);
```
Comment (stsoe) [inline src/xma/src/xmaapi/xma_utils.cpp:703]: Will this compile?  I think it should be` const XmaBufferObjPrivate*` now that argument correctly is const.

Comment (vboggara-xilinx) [reply src/xma/src/xmaapi/xma_utils.cpp:703]: not seeing any compilation issue.

**Comment by vboggara-xilinx** [comment]
@stsoe , @chvamshi-xilinx , Build is passing now. please approve.

**Review by stsoe — APPROVED** [review]

**Review by chvamshi-xilinx — APPROVED** [review]

**Comment by vboggara-xilinx** [comment]
@sarab96 , please approve.

**Review by sarab96 — APPROVED** [review]
Looks good. I'm assuming testcases are run to check the changes.

**Pull request merged**